src/python/pants/engine/scheduler.py

```diff
@@ -506,30 +508,56 @@ def product_request(self, product, subjects):
     :returns: A list of the requested products, with length match len(subjects).
     """
     request = None
+    raised_exception = None
     try:
       request = self.execution_request([product], subjects)
-    except Exception:
-      internal_errors = self._scheduler._native.cffi_extern_method_runtime_exceptions()
-      if not internal_errors:
+    except:
+      # If there are any exceptions during CFFI extern method calls, we want to return an error with
+      # them and whatever failure results from it. This typically results from unhashable types.
+      if self._scheduler._native.cffi_extern_method_runtime_exceptions():
+        raised_exception = sys.exc_info()[0:3]
+      else:
+        # Otherwise, this is likely an exception coming from somewhere else, and we don't want to
+        # swallow that, so re-raise.
         raise
 
+    # We still want to raise whenever there are any exceptions in any CFFI extern methods, even if
+    # that didn't lead to an exception in generating the execution request for some reason, so we
+    # check the extern exceptions list again.
```

Comment:
We could get rid of a bunch of the repeated conditional checks, and make this read a bit better, by just structuring this `except` block as:

```python
except:
  internal_errors = self._scheduler._native.cffi_extern_method_runtime_exceptions()
  if not internal_errors:
    raise
  error_tracebacks = [...]
  exc_type, exc_value, tv = sys.exc_info()
  ...
```

Reply:
So there's no guarantee that the `.execution_request()` method will raise an error just because the CFFI externs do -- that is currently the case, because our new `onerror` handler will make them return `None` when an exception is raised, and the current python and rust code that runs in the `.execution_request()` method will end up raising an error if those extern methods return `None`, but that's not behavior I want to depend on, hence checking outside of the `except` as well.

Reply:
And I can't see how to merge that logic with e.g. generating the `error_tracebacks` within the `except`.

Reply:
Ohhhhh I totally didn't see the fact that this is less indented than the block above! Apologies!